geth/peers/topicpool.go

```diff
@@ -112,40 +171,53 @@ func (t *TopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
 func (t *TopicPool) ConfirmDropped(server *p2p.Server, nodeID discover.NodeID) bool {
 	t.mu.Lock()
 	defer t.mu.Unlock()
+
+	discV5NodeID := discv5.NodeID(nodeID)
+
 	// either inbound or connected from another topic
-	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	peer, exist := t.connectedPeers[discV5NodeID]
 	if !exist {
 		return false
 	}
-	log.Debug("disconnect", "ID", nodeID)
-	if peer.requested {
+
+	log.Debug("disconnect", "ID", nodeID, "dismissed", peer.dismissed)
+
+	delete(t.connectedPeers, discV5NodeID)
+
+	if peer.dismissed {
+		t.addToPeerPool(peer)
 		return false
 	}
-	if t.SearchRunning() && t.connected == t.limits[0] {
-		t.period <- t.fastSync
-	}
-	t.connected--
-	t.removePeer(server, peer)
-	delete(t.peers, discv5.NodeID(nodeID))
+
+	// remove from cache only if the peer dropped by itself
 	if t.cache != nil {
-		if err := t.cache.RemovePeer(discv5.NodeID(nodeID), t.topic); err != nil {
+		if err := t.cache.RemovePeer(discV5NodeID, t.topic); err != nil {
 			log.Error("failed to remove peer from cache", "error", err)
 		}
 	}
+
+	if t.SearchRunning() && len(t.connectedPeers) < t.limits[0] {
+		t.period <- t.fastSync
+	}
+
 	return true
 }
 
 // AddPeerFromTable checks if there is a valid peer in local table and adds it to a server.
 func (t *TopicPool) AddPeerFromTable(server *p2p.Server) *discv5.Node {
 	t.mu.RLock()
 	defer t.mu.RUnlock()
-	// TODO use a heap queue and always get a peer that was discovered recently
-	for _, peer := range t.peers {
-		if !peer.connected && mclock.Now() < peer.discoveredTime+mclock.AbsTime(expirationPeriod) {
-			t.addPeer(server, peer)
-			return peer.node
-		}
+
+	// The most recently added peer is removed from the queue.
+	// If it did not expire yet, it will be added to the server.
+	// TODO(adam): investigate if it's worth to keep the peer in the queue
+	// until the server confirms it is added and in the meanwhile only adjust its priority.
+	peer := t.popFromPeerPool()
+	if peer != nil && mclock.Now() < peer.discoveredTime+mclock.AbsTime(expirationPeriod) {
```

Comment:
We clean up old peers here now.

Reply:
Also, in such a case, we can remove all pending peers from the pool because all other peers were discovered earlier so expired "even more" :D